server/trillian_log_server/main.go

```diff
@@ -28,100 +28,69 @@ import (
-	httpPortFlag        = flag.Int("http_port", 8091, "Port to serve HTTP metrics on")
+	httpPortFlag        = flag.Int("http_port", 8091, "Port to serve HTTP metrics and REST requests on (negative means disabled)")
 	dumpMetricsInterval = flag.Duration("dump_metrics_interval", 0, "If greater than 0, how often to dump metrics to the logs.")
 )
 
-func startRPCServer(registry extension.Registry) (*grpc.Server, error) {
-	// Create and publish the RPC stats objects
-	statsInterceptor := monitoring.NewRPCStatsInterceptor(util.SystemTimeSource{}, "ct", "example")
-	statsInterceptor.Publish()
-
-	// Create the server, using the interceptor to record stats on the requests
-	grpcServer := grpc.NewServer(grpc.UnaryInterceptor(statsInterceptor.Interceptor()))
-
-	logServer := server.NewTrillianLogRPCServer(registry, new(util.SystemTimeSource))
-	if err := logServer.IsHealthy(); err != nil {
-		return nil, err
-	}
-	trillian.RegisterTrillianLogServer(grpcServer, logServer)
-
-	adminServer := admin.New(registry)
-	trillian.RegisterTrillianAdminServer(grpcServer, adminServer)
-
-	reflection.Register(grpcServer)
-	return grpcServer, nil
-}
-
 func main() {
 	flag.Parse()
-	glog.CopyStandardLogTo("WARNING")
-	glog.Info("**** Log RPC Server Starting ****")
-
-	// Enable dumping of metrics to the log at regular interval,
-	// if requested.
-	if *dumpMetricsInterval > 0 {
-		go metric.DumpToLog(context.Background(), *dumpMetricsInterval)
-	}
 
 	// First make sure we can access the database, quit if not
 	db, err := mysql.OpenDB(*mySQLURI)
 	if err != nil {
-		glog.Exitf("Failed to open MySQL database: %v", err)
+		glog.Exitf("Failed to open database: %v", err)
 	}
-	defer db.Close()
+	// No defer: database ownership is delegated to server.Main
 
 	registry := extension.Registry{
 		AdminStorage:  mysql.NewAdminStorage(db),
 		SignerFactory: keys.PEMSignerFactory{},
 		LogStorage:    mysql.NewLogStorage(db),
 	}
 
-	// Start HTTP server (optional)
-	if *exportRPCMetrics {
-		glog.Infof("Creating HTP server starting on port: %d", *httpPortFlag)
-		if err := util.StartHTTPServer(*httpPortFlag); err != nil {
-			glog.Exitf("Failed to start http server on port %d: %v", *httpPortFlag, err)
-		}
-	}
+	ts := util.SystemTimeSource{}
+	stats := monitoring.NewRPCStatsInterceptor(ts, "ct", "example")
+	stats.Publish()
+	s := grpc.NewServer(grpc.UnaryInterceptor(stats.Interceptor()))
+	// No defer: server ownership is delegated to server.Main
 
-	// Set up the listener for the server
-	// TODO(Martin2112): More flexible listen address configuration
-	glog.Infof("Creating RPC server starting on port: %d", *serverPortFlag)
-	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", *serverPortFlag))
-	if err != nil {
-		glog.Exitf("Failed to listen on the server port: %d, because: %v", *serverPortFlag, err)
+	httpEndpoint := ""
+	if *httpPortFlag >= 0 {
+		httpEndpoint = fmt.Sprintf("localhost:%v", *httpPortFlag)
```

Comment:
You're probably going to want a way to specify the address to bind to too, I expect?

Operators may want to expose the RPC and HTTP servers on different addresses too, depending on the deployment specifics, would it make sense to dump the *Port flags and just have the bind address string passed in by the flags for this and the RPC endpoints?